### PR TITLE
Feature branch testing methods able to merge

### PR DESCRIPTION
**Pull request by Luzaks**
Feature branch where enumerable methods were built.

### CODE REVIEW

File: enumerable_method.rb [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    while i <= length - 1
+      yield(self[i])
+      i += 1
+    end
+    self
+  end
+
+  # each_with_index method
+
+  def each_with_index
```
Comment (shshamim63) [inline enumerable_method.rb]: Function name should be my_each_with_index. Cause the original method has the same name that you have used.

File: enumerable_method.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  def my_all?
+    arr = self
+    arr.my_each do |x|
+      return true unless block_given?
+
+      item = yield(x)
+      return false unless item
+    end
+    true
+  end
```
Comment (shshamim63) [inline enumerable_method.rb]: my_all? should cover the task below :point_down: 
1. returns true when none of the collection members are false or nil.
`[nil, true, 9].my_all?` should return false.
2. Should take an argument. If the argument is given, returns true if all of the collection is a member of such class.
`[1, 2i, 3.14].my_all?(Numeric) ` should return true.
3. When a Regex is passed as an argument, returns true if all of the collection matches the Regex
`%w[ant bear cat].my_all?(/t/)` should return false
4. When a pattern other than Regex or a Class is given, returns true if all of the collection matches the pattern
`[3, 3, 3].my_all(3)?` should return true. 
my_all?

File: enumerable_method.rb [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  def my_any?
+    p arr = self
+    arr.my_each do |x|
+      return true unless block_given?
+
+      item = yield(x)
+      return true if item
+    end
+    false
+  end
```
Comment (shshamim63) [inline enumerable_method.rb]: It should also contain all the functionality of my_all? function that has been mention above except it should work for one item.
my_any?

File: enumerable_method.rb [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  def my_none?
+    p arr = self
+    arr.my_each do |x|
+      return true unless block_given?
+
+      item = yield(x)
+      return false if item
+    end
+    true
+  end
```
Comment (shshamim63) [inline enumerable_method.rb]: It should also contain all the functionality of my_all?. That has been mention above except it should work for if none of the item matches with the criteria.
my_none?

File: enumerable_method.rb [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  # my_count method
+  def my_count
+    p arr = self
+    elements = 0
+    elements += 1 while elements < arr.length
+    elements
+  end
```
Comment (shshamim63) [inline enumerable_method.rb]: 1. Counts the number of items in enum that are equal to item if an argument is given. 
`[1, 2, 4, 2].my_count(2)` should return 2.
2. Counts the number of elements yielding a true value if a block is given.
`[1, 2, 4, 2].my_count{ |x| x%2==0 }` should return 3.
my_count

File: enumerable_method.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  # multiply_els
+
+  def multiply_els
+    arr = self
+    result = 0
+    arr.my_inject do |x, y|
+      result = yield(x, y)
+    end
+    result
+  end
```
Comment (shshamim63) [inline enumerable_method.rb]: This should be outside of the enumerable class. Cause we are checking, that our multiply_els if working properly by using the enumerable my_inject method or not.

File: enumerable_method.rb [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+    arr = self
+    map_arr = []
+    if block_given?
+      arr.my_each { |x| map_arr.push(yield(arr[x])) }
+    else
+      arr.my_each { |x| map_arr << my_proc.call(arr[x]) }
+    end
+    map_arr
+  end
```
Comment (shshamim63) [inline enumerable_method.rb]: I tried to run `(1..4).my_map { |i| i*i }` and the output should provide [1, 4, 9, 16]. But i am getting an error.

**Review by shshamim63 — COMMENTED** [review]
## Status :red_circle:  [Required Changes] :red_circle:
___
Hi @Luzaks :wave: ,
Great initiative :clap: . You have covered some method properly and some partially. So, to declare this project as complete you have to convert those partially correct methods to fully functional one.
___
### Please follow the instruction below :point_down: to complete the project.
___
I am following this [link](https://ruby-doc.org/core-2.6.5/Enumerable.html). Where I have found the description of those methods and it was mentioned on the pathwright.
pathwright
___
## :spades:  [Mandatory Changes] :spades: 
___
Follow the comments below. And all of them are the mandatory changes. 
___
#### Please Fix the mentioned part and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Slack](Shakhawat Hossain)

**Comment by shshamim63** [comment]
## Status :red_circle: [Required Changes] :red_circle: 
___
Hi @Luzaks :wave.
Awesome job :clap: , you are almost there. Just some edge cases need to cover to complete this project. I am sure that you will cover them all.
___
##### Please follow the instructions below :point_down: to complete this project.
___
### :spades: [Mandatory Changes] :spades: 
___
#### 1.  my_all? 
There is still a part that my all is not covering. That is when is try `[1, true, 'hi', []].my_all?` it should return me true. cause none of the items are false/nil.
https://github.com/Luzaks/Blocks-Enumerable-Methods/blob/299f8fe34a64bb8db0536bed2210279daebf0f29/enumerable_method.rb#L46-L56
In your code you have defined that 
```
elsif !block_given? && arg.nil?
      return false
end
```
Instead of it I have tried
```
elsif !block_given? && arg.nil?
      arr.my_each { |item| return false unless item }
end
```
and it worked.
####  2. my_any?
Returns true if at least one of the collection is not false or nil. `[nil, false, true, []].my_any?` should return true. Think about the upper hint i have provided.
https://github.com/Luzaks/Blocks-Enumerable-Methods/blob/299f8fe34a64bb8db0536bed2210279daebf0f29/enumerable_method.rb#L60-L68
3. #### my_none?
https://github.com/Luzaks/Blocks-Enumerable-Methods/blob/299f8fe34a64bb8db0536bed2210279daebf0f29/enumerable_method.rb#L72-L80
..1. I could not find the declaration of input in the code. I think it should be arg
..2. Returns true only if none of the collection members is true. If `[nil, false, nil, false].my_none?` executes, should return true. And  `[nil, false, true, []].my_none?` should return false.
4. #### multiply_els
https://github.com/Luzaks/Blocks-Enumerable-Methods/blob/299f8fe34a64bb8db0536bed2210279daebf0f29/enumerable_method.rb#L138-L145
`multiply_els` is a method which uses inject to multiply the element of an array. The example on the odinProject is something like this `multiply_els([2,4,5])`.
multiply
I think it should be like->
```
def multiply_els(arr)
      /* required part */
end
```
___
### :clubs: [Optional Changes] :clubs: 
___
1. #### my_inject
Inject method now takes two argument one is the initial and another is symbol.
If i execute `[1, 2, 3].inject(1, :+)` This will return me 7. 
___
#### Please fix the issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/) | [Slack](Shakhawat Hossain)

File: enumerable_method.rb [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    arr = self
+    if block_given? && arg.nil?
+      arr.my_each { |item| return false unless item == true }
+    elsif arg
+      arr.my_each { |item| return false unless clases(item, arg) }
+    elsif !block_given? && arg.nil?
+      arr.my_each { |item| return false unless item }
+    end
+    true
+  end
```
Comment (wathigo) [inline enumerable_method.rb]: **[REQUIRED]**
* Returns true if the block never returns false or nil.
```
Failure/Error: expect(array.my_all?(&true_block)).to eq(array.all?(&true_block))
     
       expected: true
            got: false
     
       (compared using ==)
```
* Returns true if the block ever returns a value other than false or nil.
```
Failure/Error: expect(array.my_any?(&true_block)).to eq(array.any?(&true_block))
     
       expected: true
            got: false
     
       (compared using ==)
```

File: enumerable_method.rb [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+  def my_none?(arg = nil)
+    arr = self
+    if !block_given? && arg.nil?
+      arr.my_each { |item| return false if item }
+    elsif !block_given? && arg
+      arr.my_each { |item| return false if clases(item, arg) }
+    end
+    true
+  end
```
Comment (wathigo) [inline enumerable_method.rb]: **[REQUIRED]**
* Returns true if the block never returns true for all elements.
```
Failure/Error: expect(array.my_none?(&false_block)).to eq(array.none?(&false_block))
     
       expected: false
            got: true
     
       (compared using ==)
```

File: enumerable_method.rb [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    arr = self
+    if !args.empty?
+      result = args
+    else
+      result = arr[0]
+      arr.shift
+    end
+    arr.length.times { |i| result = yield(result, arr[i]) }
+    result
+  end
```
Comment (wathigo) [inline enumerable_method.rb]: **[REQUIRED]**
* Combines all elements of enum by applying a binary operation, specified by a block.
```
Failure/Error: expect(array.my_inject(&operation)).to eq array.inject(&operation)
     
       expected: 407
            got: 415
     
       (compared using ==)
```
* When a symbol is specified combines each element of the collection by applying the symbol as a named method.
```
Failure/Error: expect(array.my_inject(:+)).to eq array.inject(:+)
     
     LocalJumpError:
       no block given (yield)
```

**Review by wathigo — CHANGES_REQUESTED** [review]
# Changes Required
## Great progress so far :clap: 
Let's fix a few issues pointed out in the comments :point_down:  before we move on.

### After implementing the changes, submit another request for review.

**Happy Coding :computer: 
Simon Wathigo**

**Review by whiz25 — APPROVED** [review]
## Amazing :clap: 
You met all the requirements for this project so it's a :+1: from me. You can go ahead and merge but before you do that, please remove any comments you have in your code.

### Martin Afani

File: enumerable_method.rb [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+    i = 0
+    while i < size
+      yield(to_a[i])
+      i += 1
+    end
+    self
+  end
+
+  # each_with_index method
```
Comment (whiz25) [inline enumerable_method.rb]: remove this comment and any you have in this file.

Comment (Luzaks) [reply enumerable_method.rb]: > remove this comment and any you have in this file.

Fixed!

**Pull request merged**